# 01 — Data Validation

## Healthcare Claims & Revenue Analytics

This notebook performs the initial structural validation of the healthcare claims datasets before any data cleaning or transformation.

### Objectives

- Verify that all expected datasets are available
- Confirm file accessibility
- Inspect dataset dimensions and column names
- Review sample records
- Identify potential structural issues before data quality assessment

> **Important:** Raw data will not be modified during validation.

In [1]:
import pandas as pd
from pathlib import Path

# --------------------------------------------------
# Project paths
# --------------------------------------------------

PROJECT_ROOT = Path.cwd().parent
DATA_DIR = PROJECT_ROOT / "00_Working_Data"

print("Project root:")
print(PROJECT_ROOT)
print("\nWorking data directory:")
print(DATA_DIR)

Project root:
c:\Users\acharya\Desktop\Healthcare_Claims_Revenue_Analytics

Working data directory:
c:\Users\acharya\Desktop\Healthcare_Claims_Revenue_Analytics\00_Working_Data


In [2]:
# --------------------------------------------------
# Dataset inventory
# --------------------------------------------------

files = {
    "beneficiary_2008": DATA_DIR / "DE1_0_2008_Beneficiary_Summary_File_Sample_1.csv",
    "inpatient": DATA_DIR / "DE1_0_2008_to_2010_Inpatient_Claims_Sample_1.csv",
    "outpatient": DATA_DIR / "DE1_0_2008_to_2010_Outpatient_Claims_Sample_1.csv",
    "beneficiary_2009": DATA_DIR / "DE1_0_2009_Beneficiary_Summary_File_Sample_1.csv",
    "beneficiary_2010": DATA_DIR / "DE1_0_2010_Beneficiary_Summary_File_Sample_1.csv",
}

print("Dataset Inventory")
print("=" * 70)

for name, file_path in files.items():
    print(f"{name:20} | {file_path.name}")

Dataset Inventory
beneficiary_2008     | DE1_0_2008_Beneficiary_Summary_File_Sample_1.csv
inpatient            | DE1_0_2008_to_2010_Inpatient_Claims_Sample_1.csv
outpatient           | DE1_0_2008_to_2010_Outpatient_Claims_Sample_1.csv
beneficiary_2009     | DE1_0_2009_Beneficiary_Summary_File_Sample_1.csv
beneficiary_2010     | DE1_0_2010_Beneficiary_Summary_File_Sample_1.csv


In [3]:
# --------------------------------------------------
# File existence and size validation
# --------------------------------------------------

print("File Validation")
print("=" * 70)

for name, file_path in files.items():

    exists = file_path.exists()

    if exists:
        size_mb = file_path.stat().st_size / (1024 * 1024)
        print(f"{name:20} | Exists: YES | Size: {size_mb:,.2f} MB")
    else:
        print(f"{name:20} | Exists: NO  | File not found")

File Validation
beneficiary_2008     | Exists: YES | Size: 13.91 MB
inpatient            | Exists: YES | Size: 15.92 MB
outpatient           | Exists: YES | Size: 154.32 MB
beneficiary_2009     | Exists: YES | Size: 13.82 MB
beneficiary_2010     | Exists: YES | Size: 13.43 MB


In [4]:
# --------------------------------------------------
# Dataset structure validation
# --------------------------------------------------

for name, file_path in files.items():

    print("\n" + "=" * 80)
    print(f"DATASET: {name}")
    print("=" * 80)

    # Read only 5 rows for structural inspection
    df_sample = pd.read_csv(file_path, nrows=5)

    print(f"Number of columns: {len(df_sample.columns)}")

    print("\nColumn names:")
    print(df_sample.columns.tolist())

    print("\nFirst 5 records:")
    display(df_sample)


DATASET: beneficiary_2008
Number of columns: 32

Column names:
['DESYNPUF_ID', 'BENE_BIRTH_DT', 'BENE_DEATH_DT', 'BENE_SEX_IDENT_CD', 'BENE_RACE_CD', 'BENE_ESRD_IND', 'SP_STATE_CODE', 'BENE_COUNTY_CD', 'BENE_HI_CVRAGE_TOT_MONS', 'BENE_SMI_CVRAGE_TOT_MONS', 'BENE_HMO_CVRAGE_TOT_MONS', 'PLAN_CVRG_MOS_NUM', 'SP_ALZHDMTA', 'SP_CHF', 'SP_CHRNKIDN', 'SP_CNCR', 'SP_COPD', 'SP_DEPRESSN', 'SP_DIABETES', 'SP_ISCHMCHT', 'SP_OSTEOPRS', 'SP_RA_OA', 'SP_STRKETIA', 'MEDREIMB_IP', 'BENRES_IP', 'PPPYMT_IP', 'MEDREIMB_OP', 'BENRES_OP', 'PPPYMT_OP', 'MEDREIMB_CAR', 'BENRES_CAR', 'PPPYMT_CAR']

First 5 records:


,DESYNPUF_ID,BENE_BIRTH_DT,BENE_DEATH_DT,BENE_SEX_IDENT_CD,BENE_RACE_CD,BENE_ESRD_IND,SP_STATE_CODE,BENE_COUNTY_CD,BENE_HI_CVRAGE_TOT_MONS,BENE_SMI_CVRAGE_TOT_MONS,...,SP_STRKETIA,MEDREIMB_IP,BENRES_IP,PPPYMT_IP,MEDREIMB_OP,BENRES_OP,PPPYMT_OP,MEDREIMB_CAR,BENRES_CAR,PPPYMT_CAR
0,00013D2EFD8E45D1,19230501,NaN,1,1,0,26,950,12,12,...,2,0.0,0.0,0.0,50.0,10.0,0.0,0.0,0.0,0.0
1,00016F745862898F,19430101,NaN,1,1,0,39,230,12,12,...,2,0.0,0.0,0.0,0.0,0.0,0.0,700.0,240.0,0.0
2,0001FDD721E223DC,19360901,NaN,2,1,0,39,280,12,12,...,2,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,00021CA6FF03E670,19410601,NaN,1,5,0,6,290,0,0,...,2,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,00024B3D2352D2D0,19360801,NaN,1,1,0,52,590,12,12,...,2,0.0,0.0,0.0,30.0,40.0,0.0,220.0,80.0,0.0



DATASET: inpatient
Number of columns: 81

Column names:
['DESYNPUF_ID', 'CLM_ID', 'SEGMENT', 'CLM_FROM_DT', 'CLM_THRU_DT', 'PRVDR_NUM', 'CLM_PMT_AMT', 'NCH_PRMRY_PYR_CLM_PD_AMT', 'AT_PHYSN_NPI', 'OP_PHYSN_NPI', 'OT_PHYSN_NPI', 'CLM_ADMSN_DT', 'ADMTNG_ICD9_DGNS_CD', 'CLM_PASS_THRU_PER_DIEM_AMT', 'NCH_BENE_IP_DDCTBL_AMT', 'NCH_BENE_PTA_COINSRNC_LBLTY_AM', 'NCH_BENE_BLOOD_DDCTBL_LBLTY_AM', 'CLM_UTLZTN_DAY_CNT', 'NCH_BENE_DSCHRG_DT', 'CLM_DRG_CD', 'ICD9_DGNS_CD_1', 'ICD9_DGNS_CD_2', 'ICD9_DGNS_CD_3', 'ICD9_DGNS_CD_4', 'ICD9_DGNS_CD_5', 'ICD9_DGNS_CD_6', 'ICD9_DGNS_CD_7', 'ICD9_DGNS_CD_8', 'ICD9_DGNS_CD_9', 'ICD9_DGNS_CD_10', 'ICD9_PRCDR_CD_1', 'ICD9_PRCDR_CD_2', 'ICD9_PRCDR_CD_3', 'ICD9_PRCDR_CD_4', 'ICD9_PRCDR_CD_5', 'ICD9_PRCDR_CD_6', 'HCPCS_CD_1', 'HCPCS_CD_2', 'HCPCS_CD_3', 'HCPCS_CD_4', 'HCPCS_CD_5', 'HCPCS_CD_6', 'HCPCS_CD_7', 'HCPCS_CD_8', 'HCPCS_CD_9', 'HCPCS_CD_10', 'HCPCS_CD_11', 'HCPCS_CD_12', 'HCPCS_CD_13', 'HCPCS_CD_14', 'HCPCS_CD_15', 'HCPCS_CD_16', 'HCPCS_CD_17', 'HCPCS_CD_

,DESYNPUF_ID,CLM_ID,SEGMENT,CLM_FROM_DT,CLM_THRU_DT,PRVDR_NUM,CLM_PMT_AMT,NCH_PRMRY_PYR_CLM_PD_AMT,AT_PHYSN_NPI,OP_PHYSN_NPI,...,HCPCS_CD_36,HCPCS_CD_37,HCPCS_CD_38,HCPCS_CD_39,HCPCS_CD_40,HCPCS_CD_41,HCPCS_CD_42,HCPCS_CD_43,HCPCS_CD_44,HCPCS_CD_45
0,00013D2EFD8E45D1,196661176988405,1,20100312,20100313,2600GD,4000.0,0.0,3139083564,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,00016F745862898F,196201177000368,1,20090412,20090418,3900MB,26000.0,0.0,6476809087,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,00016F745862898F,196661177015632,1,20090831,20090902,3900HM,5000.0,0.0,611998537,6.119985e+08,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,00016F745862898F,196091176981058,1,20090917,20090920,3913XU,5000.0,0.0,4971602784,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,00016F745862898F,196261176983265,1,20100626,20100701,3900MB,16000.0,0.0,6408400473,1.960860e+09,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN



DATASET: outpatient
Number of columns: 76

Column names:
['DESYNPUF_ID', 'CLM_ID', 'SEGMENT', 'CLM_FROM_DT', 'CLM_THRU_DT', 'PRVDR_NUM', 'CLM_PMT_AMT', 'NCH_PRMRY_PYR_CLM_PD_AMT', 'AT_PHYSN_NPI', 'OP_PHYSN_NPI', 'OT_PHYSN_NPI', 'NCH_BENE_BLOOD_DDCTBL_LBLTY_AM', 'ICD9_DGNS_CD_1', 'ICD9_DGNS_CD_2', 'ICD9_DGNS_CD_3', 'ICD9_DGNS_CD_4', 'ICD9_DGNS_CD_5', 'ICD9_DGNS_CD_6', 'ICD9_DGNS_CD_7', 'ICD9_DGNS_CD_8', 'ICD9_DGNS_CD_9', 'ICD9_DGNS_CD_10', 'ICD9_PRCDR_CD_1', 'ICD9_PRCDR_CD_2', 'ICD9_PRCDR_CD_3', 'ICD9_PRCDR_CD_4', 'ICD9_PRCDR_CD_5', 'ICD9_PRCDR_CD_6', 'NCH_BENE_PTB_DDCTBL_AMT', 'NCH_BENE_PTB_COINSRNC_AMT', 'ADMTNG_ICD9_DGNS_CD', 'HCPCS_CD_1', 'HCPCS_CD_2', 'HCPCS_CD_3', 'HCPCS_CD_4', 'HCPCS_CD_5', 'HCPCS_CD_6', 'HCPCS_CD_7', 'HCPCS_CD_8', 'HCPCS_CD_9', 'HCPCS_CD_10', 'HCPCS_CD_11', 'HCPCS_CD_12', 'HCPCS_CD_13', 'HCPCS_CD_14', 'HCPCS_CD_15', 'HCPCS_CD_16', 'HCPCS_CD_17', 'HCPCS_CD_18', 'HCPCS_CD_19', 'HCPCS_CD_20', 'HCPCS_CD_21', 'HCPCS_CD_22', 'HCPCS_CD_23', 'HCPCS_CD_24', 'HCPCS_CD_25

,DESYNPUF_ID,CLM_ID,SEGMENT,CLM_FROM_DT,CLM_THRU_DT,PRVDR_NUM,CLM_PMT_AMT,NCH_PRMRY_PYR_CLM_PD_AMT,AT_PHYSN_NPI,OP_PHYSN_NPI,...,HCPCS_CD_36,HCPCS_CD_37,HCPCS_CD_38,HCPCS_CD_39,HCPCS_CD_40,HCPCS_CD_41,HCPCS_CD_42,HCPCS_CD_43,HCPCS_CD_44,HCPCS_CD_45
0,00013D2EFD8E45D1,542192281063886,1,20080904,20080904,2600RA,50.0,0.0,4824842417,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,00016F745862898F,542272281166593,1,20090602,20090602,3901GS,30.0,0.0,2963419753,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,00016F745862898F,542282281644416,1,20090623,20090623,3939PG,30.0,0.0,5737807753,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,0001FDD721E223DC,542642281250669,1,20091011,20091011,3902NU,30.0,0.0,1233847710,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,00024B3D2352D2D0,542242281386963,1,20080712,20080712,5200TV,30.0,0.0,9688809345,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN



DATASET: beneficiary_2009
Number of columns: 32

Column names:
['DESYNPUF_ID', 'BENE_BIRTH_DT', 'BENE_DEATH_DT', 'BENE_SEX_IDENT_CD', 'BENE_RACE_CD', 'BENE_ESRD_IND', 'SP_STATE_CODE', 'BENE_COUNTY_CD', 'BENE_HI_CVRAGE_TOT_MONS', 'BENE_SMI_CVRAGE_TOT_MONS', 'BENE_HMO_CVRAGE_TOT_MONS', 'PLAN_CVRG_MOS_NUM', 'SP_ALZHDMTA', 'SP_CHF', 'SP_CHRNKIDN', 'SP_CNCR', 'SP_COPD', 'SP_DEPRESSN', 'SP_DIABETES', 'SP_ISCHMCHT', 'SP_OSTEOPRS', 'SP_RA_OA', 'SP_STRKETIA', 'MEDREIMB_IP', 'BENRES_IP', 'PPPYMT_IP', 'MEDREIMB_OP', 'BENRES_OP', 'PPPYMT_OP', 'MEDREIMB_CAR', 'BENRES_CAR', 'PPPYMT_CAR']

First 5 records:


,DESYNPUF_ID,BENE_BIRTH_DT,BENE_DEATH_DT,BENE_SEX_IDENT_CD,BENE_RACE_CD,BENE_ESRD_IND,SP_STATE_CODE,BENE_COUNTY_CD,BENE_HI_CVRAGE_TOT_MONS,BENE_SMI_CVRAGE_TOT_MONS,...,SP_STRKETIA,MEDREIMB_IP,BENRES_IP,PPPYMT_IP,MEDREIMB_OP,BENRES_OP,PPPYMT_OP,MEDREIMB_CAR,BENRES_CAR,PPPYMT_CAR
0,00013D2EFD8E45D1,19230501,NaN,1,1,0,26,950,12,12,...,2,0.0,0.0,0.0,0.0,0.0,0.0,100.0,20.0,0.0
1,00016F745862898F,19430101,NaN,1,1,0,39,230,12,12,...,1,36000.0,3204.0,0.0,60.0,70.0,0.0,1350.0,530.0,100.0
2,0001FDD721E223DC,19360901,NaN,2,1,0,39,280,12,12,...,2,0.0,0.0,0.0,30.0,50.0,0.0,20.0,0.0,0.0
3,00021CA6FF03E670,19410601,NaN,1,5,0,6,290,12,12,...,2,0.0,0.0,0.0,0.0,0.0,0.0,90.0,10.0,0.0
4,00024B3D2352D2D0,19360801,NaN,1,1,0,52,590,12,12,...,2,0.0,0.0,0.0,90.0,40.0,0.0,410.0,140.0,0.0



DATASET: beneficiary_2010
Number of columns: 32

Column names:
['DESYNPUF_ID', 'BENE_BIRTH_DT', 'BENE_DEATH_DT', 'BENE_SEX_IDENT_CD', 'BENE_RACE_CD', 'BENE_ESRD_IND', 'SP_STATE_CODE', 'BENE_COUNTY_CD', 'BENE_HI_CVRAGE_TOT_MONS', 'BENE_SMI_CVRAGE_TOT_MONS', 'BENE_HMO_CVRAGE_TOT_MONS', 'PLAN_CVRG_MOS_NUM', 'SP_ALZHDMTA', 'SP_CHF', 'SP_CHRNKIDN', 'SP_CNCR', 'SP_COPD', 'SP_DEPRESSN', 'SP_DIABETES', 'SP_ISCHMCHT', 'SP_OSTEOPRS', 'SP_RA_OA', 'SP_STRKETIA', 'MEDREIMB_IP', 'BENRES_IP', 'PPPYMT_IP', 'MEDREIMB_OP', 'BENRES_OP', 'PPPYMT_OP', 'MEDREIMB_CAR', 'BENRES_CAR', 'PPPYMT_CAR']

First 5 records:


,DESYNPUF_ID,BENE_BIRTH_DT,BENE_DEATH_DT,BENE_SEX_IDENT_CD,BENE_RACE_CD,BENE_ESRD_IND,SP_STATE_CODE,BENE_COUNTY_CD,BENE_HI_CVRAGE_TOT_MONS,BENE_SMI_CVRAGE_TOT_MONS,...,SP_STRKETIA,MEDREIMB_IP,BENRES_IP,PPPYMT_IP,MEDREIMB_OP,BENRES_OP,PPPYMT_OP,MEDREIMB_CAR,BENRES_CAR,PPPYMT_CAR
0,00013D2EFD8E45D1,19230501,NaN,1,1,0,26,950,12,12,...,2,4000.0,1100.0,0.0,0.0,0.0,0.0,90.0,30.0,0.0
1,00016F745862898F,19430101,NaN,1,1,Y,39,230,12,12,...,2,16000.0,1100.0,0.0,0.0,0.0,0.0,930.0,150.0,0.0
2,0001FDD721E223DC,19360901,NaN,2,1,0,39,280,12,12,...,2,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,00021CA6FF03E670,19410601,NaN,1,5,0,6,290,12,12,...,2,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,00024B3D2352D2D0,19360801,NaN,1,1,0,52,590,9,12,...,2,0.0,0.0,0.0,40.0,0.0,0.0,590.0,60.0,0.0


## Initial Validation Findings

The initial structural validation confirms that all five expected datasets are available and readable.

### Dataset Structure

- Beneficiary Summary 2008 contains 32 columns.
- Inpatient Claims contains 81 columns.
- Outpatient Claims contains 76 columns.
- Beneficiary Summary 2009 contains 32 columns.
- Beneficiary Summary 2010 contains 32 columns.

### Key Identifiers

- `DESYNPUF_ID` is present across the beneficiary and claims datasets and can be used to relate beneficiaries to claims.
- `CLM_ID` is available in the inpatient and outpatient claims datasets and can be investigated as a claim-level identifier.

### Key Analytical Fields

The claims datasets contain:

- Claim payment amounts (`CLM_PMT_AMT`)
- Provider identifiers (`PRVDR_NUM`)
- Physician identifiers
- Diagnosis codes
- Procedure codes
- HCPCS codes

The inpatient dataset also contains utilization information such as inpatient days.

### Important Observation

A dedicated claim denial status field is not apparent in the initial schema inspection. Therefore, denial-rate analysis should not be assumed to be directly available from this dataset and will be evaluated during the subsequent data-quality and field-definition assessment.

### Data Handling Rule

No source CSV files were modified during validation. The datasets were inspected in read-only mode.

In [5]:
# --------------------------------------------------
# Record count validation
# --------------------------------------------------

print("Record Count Validation")
print("=" * 70)

record_counts = {}

for name, file_path in files.items():

    # Count rows without loading the complete dataset into memory
    row_count = sum(1 for _ in open(file_path, encoding="utf-8")) - 1

    record_counts[name] = row_count

    print(f"{name:20} | {row_count:,} records")

Record Count Validation
beneficiary_2008     | 116,352 records
inpatient            | 66,773 records
outpatient           | 790,790 records
beneficiary_2009     | 114,538 records
beneficiary_2010     | 112,754 records


In [6]:
# --------------------------------------------------
# Duplicate and identifier validation
# --------------------------------------------------

print("Duplicate and Identifier Validation")
print("=" * 70)

# --------------------------------------------------
# Beneficiary datasets
# --------------------------------------------------

beneficiary_files = {
    "beneficiary_2008": files["beneficiary_2008"],
    "beneficiary_2009": files["beneficiary_2009"],
    "beneficiary_2010": files["beneficiary_2010"],
}

for name, file_path in beneficiary_files.items():

    df = pd.read_csv(file_path, usecols=["DESYNPUF_ID"])

    total_rows = len(df)
    unique_ids = df["DESYNPUF_ID"].nunique()
    duplicate_rows = total_rows - unique_ids

    print(f"\n{name}")
    print(f"Total records       : {total_rows:,}")
    print(f"Unique beneficiary IDs: {unique_ids:,}")
    print(f"Duplicate ID rows   : {duplicate_rows:,}")


# --------------------------------------------------
# Claims datasets
# --------------------------------------------------

claim_files = {
    "inpatient": files["inpatient"],
    "outpatient": files["outpatient"],
}

for name, file_path in claim_files.items():

    unique_ids = set()
    duplicate_ids = set()
    total_rows = 0

    for chunk in pd.read_csv(
        file_path,
        usecols=["CLM_ID"],
        chunksize=100_000
    ):

        for claim_id in chunk["CLM_ID"].dropna():

            total_rows += 1

            if claim_id in unique_ids:
                duplicate_ids.add(claim_id)
            else:
                unique_ids.add(claim_id)

    print(f"\n{name}")
    print(f"Total claim records : {total_rows:,}")
    print(f"Unique claim IDs    : {len(unique_ids):,}")
    print(f"Duplicate claim IDs : {len(duplicate_ids):,}")

Duplicate and Identifier Validation

beneficiary_2008
Total records       : 116,352
Unique beneficiary IDs: 116,352
Duplicate ID rows   : 0

beneficiary_2009
Total records       : 114,538
Unique beneficiary IDs: 114,538
Duplicate ID rows   : 0

beneficiary_2010
Total records       : 112,754
Unique beneficiary IDs: 112,754
Duplicate ID rows   : 0

inpatient
Total claim records : 66,773
Unique claim IDs    : 66,705
Duplicate claim IDs : 68

outpatient
Total claim records : 790,790
Unique claim IDs    : 779,815
Duplicate claim IDs : 10,975


## Identifier Validation Findings

The beneficiary summary files contain unique `DESYNPUF_ID` values within each annual dataset.

### Beneficiary Files

- 2008: 116,352 records and 116,352 unique beneficiary IDs
- 2009: 114,538 records and 114,538 unique beneficiary IDs
- 2010: 112,754 records and 112,754 unique beneficiary IDs

No duplicate beneficiary IDs were identified within the individual annual files.

### Claims Files

Repeated `CLM_ID` values were identified:

- Inpatient: 68 duplicate claim-ID occurrences
- Outpatient: 10,975 duplicate claim-ID occurrences

These repeated claim IDs are not automatically classified as erroneous duplicates. The claims datasets contain a `SEGMENT` field, so the relationship between `CLM_ID` and `SEGMENT` must be investigated before any records are removed.

### Data Quality Decision

No records will be deleted at this stage.

The repeated claim IDs will be investigated during the detailed data-quality assessment to determine whether they represent:

- multiple segments belonging to the same claim,
- legitimate repeated claim records, or
- actual duplicate records.

In [7]:
# --------------------------------------------------
# Investigate CLM_ID + SEGMENT uniqueness
# --------------------------------------------------

print("CLM_ID + SEGMENT Validation")
print("=" * 70)

for name, file_path in claim_files.items():

    pair_counts = {}

    for chunk in pd.read_csv(
        file_path,
        usecols=["CLM_ID", "SEGMENT"],
        chunksize=100_000
    ):
        chunk = chunk.dropna(subset=["CLM_ID"])

        for claim_id, segment in zip(
            chunk["CLM_ID"],
            chunk["SEGMENT"]
        ):
            key = (claim_id, segment)
            pair_counts[key] = pair_counts.get(key, 0) + 1

    total_records = sum(pair_counts.values())
    unique_pairs = len(pair_counts)

    duplicate_pairs = sum(
        count > 1 for count in pair_counts.values()
    )

    duplicate_rows = sum(
        count - 1
        for count in pair_counts.values()
        if count > 1
    )

    print(f"\n{name}")
    print(f"Total claim records       : {total_records:,}")
    print(f"Unique CLM_ID + SEGMENT  : {unique_pairs:,}")
    print(f"Duplicate pairs           : {duplicate_pairs:,}")
    print(f"Duplicate rows            : {duplicate_rows:,}")

CLM_ID + SEGMENT Validation

inpatient
Total claim records       : 66,773
Unique CLM_ID + SEGMENT  : 66,773
Duplicate pairs           : 0
Duplicate rows            : 0

outpatient
Total claim records       : 790,790
Unique CLM_ID + SEGMENT  : 790,790
Duplicate pairs           : 0
Duplicate rows            : 0


## CLM_ID + SEGMENT Validation Findings

The repeated `CLM_ID` values identified in the initial identifier validation were investigated using the combined `CLM_ID + SEGMENT` key.

### Results

- Inpatient Claims:
  - 66,773 claim records
  - 66,773 unique `CLM_ID + SEGMENT` combinations
  - 0 duplicate combinations

- Outpatient Claims:
  - 790,790 claim records
  - 790,790 unique `CLM_ID + SEGMENT` combinations
  - 0 duplicate combinations

### Interpretation

Although some `CLM_ID` values occur more than once, each `CLM_ID + SEGMENT` combination is unique.

This indicates that the repeated `CLM_ID` values are associated with different claim segments rather than being exact duplicate records.

### Data Quality Decision

The repeated `CLM_ID` values will **not** be treated as erroneous duplicates.

The composite key `CLM_ID + SEGMENT` should be considered when performing claim-level record validation and analysis.

No records will be removed based on the repeated `CLM_ID` values.

## Missing Value Validation

The next validation step assesses the completeness of the datasets by identifying missing values in each field.

Missing values are not automatically treated as errors. In healthcare claims data, some fields may legitimately be blank depending on the claim type, beneficiary status, or whether a particular field applies to a record.

The purpose of this step is to identify fields requiring further investigation during the Data Quality Assessment phase.

> **Important:** No missing values will be filled, removed, or otherwise modified during this validation step.

In [8]:
# --------------------------------------------------
# Missing Value Validation
# --------------------------------------------------

print("Missing Value Validation")
print("=" * 70)

for name, file_path in files.items():

    print(f"\n{name}")
    print("-" * 70)

    missing_counts = {}

    for chunk in pd.read_csv(
        file_path,
        chunksize=100_000
    ):
        chunk_missing = chunk.isna().sum()

        for column, count in chunk_missing.items():
            missing_counts[column] = (
                missing_counts.get(column, 0) + count
            )

    total_records = sum(1 for _ in open(file_path, encoding="utf-8")) - 1

    missing_df = pd.DataFrame({
        "Missing_Count": missing_counts
    })

    missing_df["Missing_Percentage"] = (
        missing_df["Missing_Count"] / total_records * 100
    )

    missing_df = (
        missing_df[missing_df["Missing_Count"] > 0]
        .sort_values("Missing_Count", ascending=False)
    )

    if missing_df.empty:
        print("No missing values identified.")
    else:
        display(missing_df)

Missing Value Validation

beneficiary_2008
----------------------------------------------------------------------


,Missing_Count,Missing_Percentage
BENE_DEATH_DT,114538,98.440938



inpatient
----------------------------------------------------------------------


,Missing_Count,Missing_Percentage
HCPCS_CD_9,66773,100.000000
HCPCS_CD_13,66773,100.000000
HCPCS_CD_8,66773,100.000000
HCPCS_CD_7,66773,100.000000
HCPCS_CD_6,66773,100.000000
...,...,...
ICD9_DGNS_CD_2,526,0.787744
ICD9_DGNS_CD_1,95,0.142273
CLM_FROM_DT,68,0.101838
CLM_THRU_DT,68,0.101838



outpatient
----------------------------------------------------------------------


C:\Users\acharya\AppData\Local\Temp\ipykernel_22412\1806222658.py:15: DtypeWarning: Columns (0: ICD9_PRCDR_CD_2, 1: ICD9_PRCDR_CD_3, 2: ICD9_PRCDR_CD_4, 3: ICD9_PRCDR_CD_5, 4: ICD9_PRCDR_CD_6) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv(
C:\Users\acharya\AppData\Local\Temp\ipykernel_22412\1806222658.py:15: DtypeWarning: Columns (0: ICD9_PRCDR_CD_2, 1: ICD9_PRCDR_CD_3) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv(
C:\Users\acharya\AppData\Local\Temp\ipykernel_22412\1806222658.py:15: DtypeWarning: Columns (0: ICD9_DGNS_CD_10, 1: ICD9_PRCDR_CD_2, 2: ICD9_PRCDR_CD_3, 3: ICD9_PRCDR_CD_4, 4: ICD9_PRCDR_CD_5) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv(
C:\Users\acharya\AppData\Local\Temp\ipykernel_22412\1806222658.py:15: DtypeWarning: Columns (0: ICD9_PRCDR_CD_2, 1: ICD9_PRCDR_CD_3, 2: ICD9_PRCDR_CD_5) have mixed types. Spec

,Missing_Count,Missing_Percentage
HCPCS_CD_45,790790,100.000000
ICD9_PRCDR_CD_6,790761,99.996333
ICD9_PRCDR_CD_5,790755,99.995574
ICD9_PRCDR_CD_4,790742,99.993930
ICD9_PRCDR_CD_3,790717,99.990769
...,...,...
HCPCS_CD_1,144530,18.276660
AT_PHYSN_NPI,17790,2.249649
CLM_FROM_DT,11253,1.423007
CLM_THRU_DT,11253,1.423007



beneficiary_2009
----------------------------------------------------------------------


,Missing_Count,Missing_Percentage
BENE_DEATH_DT,112754,98.442438



beneficiary_2010
----------------------------------------------------------------------


,Missing_Count,Missing_Percentage
BENE_DEATH_DT,110891,98.34773


## Missing Value Validation Findings

The completeness assessment identified different patterns of missingness across the beneficiary and claims datasets.

### Beneficiary Files

`BENE_DEATH_DT` is missing for approximately 98% of beneficiary records across the three annual beneficiary files.

Because a death date is only expected for beneficiaries with a recorded death event, this high level of missingness is not automatically considered a data-quality error.

### Inpatient Claims

Several fields contain substantial missingness, including some higher-numbered HCPCS fields.

Some claim fields, such as diagnosis codes and claim dates, have relatively low levels of missingness.

### Outpatient Claims

The outpatient dataset contains several fields with extremely high missingness.

`HCPCS_CD_45` is completely missing, while several procedure-code fields have missingness above 99%.

`HCPCS_CD_1` has approximately 18.3% missing values.

Some physician identifiers and claim dates also contain partial missingness.

### Data Type Observation

During the outpatient validation, pandas generated `DtypeWarning` messages for several ICD-9 diagnosis/procedure columns because values were detected with mixed data types.

This issue will be investigated during the data-quality assessment and cleaning stages.

### Data Quality Decision

Missing values will not be removed or imputed during the validation stage.

Fields with very high or complete missingness will be evaluated for removal during the cleaning process based on their analytical relevance and business requirements.

## Data Type and Date Validation

The next validation step examines the data types assigned to important fields and checks whether date fields can be interpreted consistently.

The assessment focuses on:

- Beneficiary identifiers
- Claim identifiers
- Provider identifiers
- Claim payment amounts
- Claim dates
- Inpatient admission and discharge dates
- Diagnosis and procedure code fields

This step identifies potential type-conversion issues before data cleaning.

> **Important:** No data types or source values will be modified during validation.

In [9]:
# --------------------------------------------------
# Data Type Validation
# --------------------------------------------------

print("Data Type Validation")
print("=" * 70)

validation_columns = {
    "beneficiary_2008": [
        "DESYNPUF_ID",
        "BENE_BIRTH_DT",
        "BENE_DEATH_DT",
        "BENE_SEX_IDENT_CD",
        "BENE_RACE_CD",
        "MEDREIMB_IP",
        "MEDREIMB_OP",
        "MEDREIMB_CAR"
    ],
    
    "inpatient": [
        "DESYNPUF_ID",
        "CLM_ID",
        "SEGMENT",
        "CLM_FROM_DT",
        "CLM_THRU_DT",
        "CLM_PMT_AMT",
        "PRVDR_NUM",
        "CLM_UTLZTN_DAY_CNT",
        "ICD9_DGNS_CD_1",
        "ICD9_PRCDR_CD_1"
    ],
    
    "outpatient": [
        "DESYNPUF_ID",
        "CLM_ID",
        "SEGMENT",
        "CLM_FROM_DT",
        "CLM_THRU_DT",
        "CLM_PMT_AMT",
        "PRVDR_NUM",
        "ICD9_DGNS_CD_1",
        "ICD9_PRCDR_CD_1",
        "HCPCS_CD_1"
    ]
}

for dataset_name, columns in validation_columns.items():

    print(f"\n{dataset_name}")
    print("-" * 70)

    file_path = files[dataset_name]

    sample = pd.read_csv(
        file_path,
        usecols=columns,
        nrows=10_000,
        low_memory=False
    )

    dtype_df = pd.DataFrame({
        "Column": sample.columns,
        "Data_Type": sample.dtypes.astype(str).values,
        "Non_Null_Count": sample.notna().sum().values
    })

    display(dtype_df)

Data Type Validation

beneficiary_2008
----------------------------------------------------------------------


,Column,Data_Type,Non_Null_Count
0,DESYNPUF_ID,str,10000
1,BENE_BIRTH_DT,int64,10000
2,BENE_DEATH_DT,float64,150
3,BENE_SEX_IDENT_CD,int64,10000
4,BENE_RACE_CD,int64,10000
5,MEDREIMB_IP,float64,10000
6,MEDREIMB_OP,float64,10000
7,MEDREIMB_CAR,float64,10000



inpatient
----------------------------------------------------------------------


,Column,Data_Type,Non_Null_Count
0,DESYNPUF_ID,str,10000
1,CLM_ID,int64,10000
2,SEGMENT,int64,10000
3,CLM_FROM_DT,float64,9988
4,CLM_THRU_DT,float64,9988
5,PRVDR_NUM,str,10000
6,CLM_PMT_AMT,float64,10000
7,CLM_UTLZTN_DAY_CNT,float64,9988
8,ICD9_DGNS_CD_1,str,9983
9,ICD9_PRCDR_CD_1,float64,5690



outpatient
----------------------------------------------------------------------


,Column,Data_Type,Non_Null_Count
0,DESYNPUF_ID,str,10000
1,CLM_ID,int64,10000
2,SEGMENT,int64,10000
3,CLM_FROM_DT,float64,9860
4,CLM_THRU_DT,float64,9860
5,PRVDR_NUM,str,10000
6,CLM_PMT_AMT,float64,10000
7,ICD9_DGNS_CD_1,str,9933
8,ICD9_PRCDR_CD_1,float64,5
9,HCPCS_CD_1,str,8271


## Data Type Validation Findings

The data-type assessment identified several fields that require explicit handling during the cleaning stage.

### Identifiers

- `DESYNPUF_ID` is stored as a string in the beneficiary and claims datasets.
- `PRVDR_NUM` is stored as a string.
- `CLM_ID` and `SEGMENT` are stored as integers.

These representations are suitable for identifier fields because identifiers should not be treated as numerical measurements.

### Financial Fields

`CLM_PMT_AMT` and the beneficiary reimbursement fields are stored as numeric values, which is appropriate for financial calculations.

### Date Fields

Claim dates such as `CLM_FROM_DT` and `CLM_THRU_DT` are currently represented as numeric/float values rather than proper datetime fields.

Beneficiary birth and death dates are also represented numerically.

These fields will require conversion to appropriate date formats during the cleaning stage.

### Diagnosis and Procedure Codes

Diagnosis and procedure code fields show inconsistent pandas data types.

For example, `ICD9_DGNS_CD_1` is interpreted as a string, while `ICD9_PRCDR_CD_1` is interpreted as `float64`.

This occurs because coded fields can contain missing values and may be interpreted differently by pandas.

### Data Quality Decision

No data types will be changed during validation.

Date conversion and explicit treatment of diagnosis, procedure, and HCPCS code fields will be performed during the data-cleaning stage after the complete quality assessment.

## Date Validity and Chronological Validation

This validation examines the date fields in the beneficiary and claims datasets.

The assessment checks:

- Whether beneficiary birth dates can be interpreted as valid dates
- Whether beneficiary death dates can be interpreted as valid dates
- Whether claim start and end dates can be interpreted as valid dates
- Whether `CLM_FROM_DT` occurs on or before `CLM_THRU_DT`

Invalid or logically inconsistent dates will be documented for investigation during the cleaning stage.

> **Important:** No date values will be modified during this validation step.

In [10]:
# --------------------------------------------------
# Date Validity and Chronological Validation
# --------------------------------------------------

print("Date Validity and Chronological Validation")
print("=" * 70)


# --------------------------------------------------
# Beneficiary Date Validation
# --------------------------------------------------

print("\nBENEFICIARY DATE VALIDATION")
print("-" * 70)

for name in [
    "beneficiary_2008",
    "beneficiary_2009",
    "beneficiary_2010"
]:

    file_path = files[name]

    df = pd.read_csv(
        file_path,
        usecols=["BENE_BIRTH_DT", "BENE_DEATH_DT"],
        low_memory=False
    )

    birth_dates = pd.to_datetime(
        df["BENE_BIRTH_DT"].astype("Int64").astype("string"),
        format="%Y%m%d",
        errors="coerce"
    )

    death_dates = pd.to_datetime(
        df["BENE_DEATH_DT"].astype("Int64").astype("string"),
        format="%Y%m%d",
        errors="coerce"
    )

    invalid_birth = (
        df["BENE_BIRTH_DT"].notna() &
        birth_dates.isna()
    ).sum()

    invalid_death = (
        df["BENE_DEATH_DT"].notna() &
        death_dates.isna()
    ).sum()

    death_before_birth = (
        death_dates.notna() &
        birth_dates.notna() &
        (death_dates < birth_dates)
    ).sum()

    print(f"\n{name}")
    print(f"Invalid birth dates       : {invalid_birth:,}")
    print(f"Invalid death dates       : {invalid_death:,}")
    print(f"Death before birth        : {death_before_birth:,}")


# --------------------------------------------------
# Claims Date Validation
# --------------------------------------------------

print("\nCLAIMS DATE VALIDATION")
print("-" * 70)

for name in ["inpatient", "outpatient"]:

    file_path = files[name]

    date_columns = [
        "CLM_FROM_DT",
        "CLM_THRU_DT"
    ]

    df = pd.read_csv(
        file_path,
        usecols=date_columns,
        low_memory=False
    )

    from_dates = pd.to_datetime(
        df["CLM_FROM_DT"].astype("Int64").astype("string"),
        format="%Y%m%d",
        errors="coerce"
    )

    thru_dates = pd.to_datetime(
        df["CLM_THRU_DT"].astype("Int64").astype("string"),
        format="%Y%m%d",
        errors="coerce"
    )

    invalid_from = (
        df["CLM_FROM_DT"].notna() &
        from_dates.isna()
    ).sum()

    invalid_thru = (
        df["CLM_THRU_DT"].notna() &
        thru_dates.isna()
    ).sum()

    from_after_thru = (
        from_dates.notna() &
        thru_dates.notna() &
        (from_dates > thru_dates)
    ).sum()

    print(f"\n{name}")
    print(f"Invalid claim start dates : {invalid_from:,}")
    print(f"Invalid claim end dates   : {invalid_thru:,}")
    print(f"Start date after end date : {from_after_thru:,}")

Date Validity and Chronological Validation

BENEFICIARY DATE VALIDATION
----------------------------------------------------------------------

beneficiary_2008
Invalid birth dates       : 0
Invalid death dates       : 0
Death before birth        : 0

beneficiary_2009
Invalid birth dates       : 0
Invalid death dates       : 0
Death before birth        : 0

beneficiary_2010
Invalid birth dates       : 0
Invalid death dates       : 0
Death before birth        : 0

CLAIMS DATE VALIDATION
----------------------------------------------------------------------

inpatient
Invalid claim start dates : 0
Invalid claim end dates   : 0
Start date after end date : 0

outpatient
Invalid claim start dates : 0
Invalid claim end dates   : 0
Start date after end date : 0


## Date Validation Findings

The date validation identified no invalid or chronologically inconsistent dates in the datasets.

### Beneficiary Dates

Across the 2008, 2009, and 2010 beneficiary files:

- No invalid birth dates were identified.
- No invalid death dates were identified among populated death-date fields.
- No beneficiary had a recorded death date earlier than the recorded birth date.

### Claims Dates

For both inpatient and outpatient claims:

- No invalid claim start dates were identified.
- No invalid claim end dates were identified.
- No records had a claim start date later than the claim end date.

### Data Quality Decision

The date fields are structurally valid based on the checks performed.

During data cleaning, the numeric date representations will be converted to proper datetime fields for analysis, filtering, time-series reporting, and Power BI use.

No date records will be removed based on the validation performed.

## Financial and Numeric Validation

This validation examines key numeric fields used for claims and financial analysis.

The assessment focuses on:

- Claim payment amounts (`CLM_PMT_AMT`)
- Primary payer amounts
- Beneficiary liability amounts
- Inpatient utilization days
- Beneficiary reimbursement amounts

The checks identify:

- Negative values
- Zero values
- Minimum and maximum values
- Basic distribution statistics

Negative values are not automatically considered errors because some financial fields may represent adjustments or offsets. Any unusual values will therefore be investigated before a cleaning decision is made.

> **Important:** No financial values will be modified during this validation step.

In [11]:
# --------------------------------------------------
# Financial and Numeric Validation
# --------------------------------------------------

print("Financial and Numeric Validation")
print("=" * 70)


# --------------------------------------------------
# Claims Financial Fields
# --------------------------------------------------

claim_numeric_columns = [
    "CLM_PMT_AMT",
    "NCH_PRMRY_PYR_CLM_PD_AMT"
]

for name in ["inpatient", "outpatient"]:

    print(f"\n{name}")
    print("-" * 70)

    file_path = files[name]

    summary = {
        column: {
            "count": 0,
            "negative": 0,
            "zero": 0,
            "minimum": None,
            "maximum": None,
            "sum": 0
        }
        for column in claim_numeric_columns
    }

    for chunk in pd.read_csv(
        file_path,
        usecols=claim_numeric_columns,
        chunksize=100_000,
        low_memory=False
    ):

        for column in claim_numeric_columns:

            series = chunk[column].dropna()

            if len(series) == 0:
                continue

            summary[column]["count"] += len(series)
            summary[column]["negative"] += (series < 0).sum()
            summary[column]["zero"] += (series == 0).sum()

            chunk_min = series.min()
            chunk_max = series.max()

            if (
                summary[column]["minimum"] is None
                or chunk_min < summary[column]["minimum"]
            ):
                summary[column]["minimum"] = chunk_min

            if (
                summary[column]["maximum"] is None
                or chunk_max > summary[column]["maximum"]
            ):
                summary[column]["maximum"] = chunk_max

            summary[column]["sum"] += series.sum()

    financial_summary = pd.DataFrame(summary).T

    display(financial_summary)


# --------------------------------------------------
# Inpatient Utilization
# --------------------------------------------------

print("\nINPATIENT UTILIZATION")
print("-" * 70)

inpatient_file = files["inpatient"]

utilization_summary = {
    "count": 0,
    "negative": 0,
    "zero": 0,
    "minimum": None,
    "maximum": None
}

for chunk in pd.read_csv(
    inpatient_file,
    usecols=["CLM_UTLZTN_DAY_CNT"],
    chunksize=100_000,
    low_memory=False
):

    series = chunk["CLM_UTLZTN_DAY_CNT"].dropna()

    if len(series) == 0:
        continue

    utilization_summary["count"] += len(series)
    utilization_summary["negative"] += (series < 0).sum()
    utilization_summary["zero"] += (series == 0).sum()

    chunk_min = series.min()
    chunk_max = series.max()

    if (
        utilization_summary["minimum"] is None
        or chunk_min < utilization_summary["minimum"]
    ):
        utilization_summary["minimum"] = chunk_min

    if (
        utilization_summary["maximum"] is None
        or chunk_max > utilization_summary["maximum"]
    ):
        utilization_summary["maximum"] = chunk_max

display(pd.DataFrame([utilization_summary]))


# --------------------------------------------------
# Beneficiary Reimbursement Fields
# --------------------------------------------------

print("\nBENEFICIARY REIMBURSEMENT FIELDS")
print("-" * 70)

beneficiary_numeric_columns = [
    "MEDREIMB_IP",
    "BENRES_IP",
    "PPPYMT_IP",
    "MEDREIMB_OP",
    "BENRES_OP",
    "PPPYMT_OP",
    "MEDREIMB_CAR",
    "BENRES_CAR",
    "PPPYMT_CAR"
]

beneficiary_file = files["beneficiary_2010"]

beneficiary_summary = {}

for column in beneficiary_numeric_columns:

    values = pd.read_csv(
        beneficiary_file,
        usecols=[column],
        low_memory=False
    )[column].dropna()

    beneficiary_summary[column] = {
        "count": len(values),
        "negative": (values < 0).sum(),
        "zero": (values == 0).sum(),
        "minimum": values.min(),
        "maximum": values.max(),
        "sum": values.sum()
    }

display(pd.DataFrame(beneficiary_summary).T)

Financial and Numeric Validation

inpatient
----------------------------------------------------------------------


,count,negative,zero,minimum,maximum,sum
CLM_PMT_AMT,66773.0,55.0,2160.0,-8000.0,57000.0,639260180.0
NCH_PRMRY_PYR_CLM_PD_AMT,66773.0,0.0,65246.0,0.0,68000.0,26635700.0



outpatient
----------------------------------------------------------------------


,count,negative,zero,minimum,maximum,sum
CLM_PMT_AMT,790790.0,2566.0,30105.0,-100.0,3300.0,224524710.0
NCH_PRMRY_PYR_CLM_PD_AMT,790790.0,0.0,783826.0,0.0,14000.0,8097500.0



INPATIENT UTILIZATION
----------------------------------------------------------------------


,count,negative,zero,minimum,maximum
0,66705,0,2266,0.0,136.0



BENEFICIARY REIMBURSEMENT FIELDS
----------------------------------------------------------------------


,count,negative,zero,minimum,maximum,sum
MEDREIMB_IP,112754.0,7.0,101023.0,-3000.0,105000.0,139989740.0
BENRES_IP,112754.0,0.0,100788.0,0.0,36200.0,16241170.0
PPPYMT_IP,112754.0,0.0,112463.0,0.0,68000.0,5090300.0
MEDREIMB_OP,112754.0,85.0,58598.0,-100.0,36800.0,48755220.0
BENRES_OP,112754.0,0.0,64941.0,0.0,9500.0,14744380.0
PPPYMT_OP,112754.0,0.0,111251.0,0.0,15000.0,1551930.0
MEDREIMB_CAR,112754.0,0.0,27765.0,0.0,10210.0,95573590.0
BENRES_CAR,112754.0,0.0,30037.0,0.0,2760.0,27085550.0
PPPYMT_CAR,112754.0,0.0,100508.0,0.0,1830.0,1450970.0


## Financial and Numeric Validation Findings

The financial validation identified both zero and negative values in several payment and reimbursement fields.

### Claims Payment Amounts

#### Inpatient Claims

`CLM_PMT_AMT` contains:

- 66,773 non-null values
- 55 negative values
- 2,160 zero values
- Minimum: -8,000
- Maximum: 57,000

#### Outpatient Claims

`CLM_PMT_AMT` contains:

- 790,790 non-null values
- 2,566 negative values
- 30,105 zero values
- Minimum: -100
- Maximum: 3,300

### Primary Payer Amounts

`NCH_PRMRY_PYR_CLM_PD_AMT` contains no negative values in either claims dataset.

However, zero values are very common, particularly in the outpatient claims data. Therefore, zero values should not automatically be interpreted as missing or erroneous.

### Inpatient Utilization

`CLM_UTLZTN_DAY_CNT` contains no negative values.

The observed range is 0 to 136 days.

### Beneficiary Reimbursement Fields

Negative values were identified in:

- `MEDREIMB_IP`: 7 records
- `MEDREIMB_OP`: 85 records

The remaining beneficiary reimbursement fields did not contain negative values.

### Data Quality Decision

Negative financial values will not be removed during validation.

They will be retained in the source data and investigated during the cleaning and analytical-definition stages.

For financial reporting, the project will distinguish between:

- **Gross/source payment values**
- **Negative or adjustment-like values**
- **Zero-payment records**

This prevents potentially meaningful source records from being removed without evidence.

Zero values will also be retained because a zero payment can represent a legitimate financial outcome rather than a missing value.

## Beneficiary-to-Claim Relationship Validation

The beneficiary and claims datasets are related through `DESYNPUF_ID`.

This validation checks whether claim records can be linked to a beneficiary present in the annual beneficiary summary files.

The assessment will identify:

- Claims with a matching beneficiary ID
- Claims whose beneficiary ID is not present in the beneficiary files
- The percentage of claims that can be linked

Because the beneficiary files are provided separately by year, the three annual beneficiary populations will first be combined logically for validation purposes.

> **Important:** No source data will be modified during this validation.

In [12]:
# --------------------------------------------------
# Beneficiary-to-Claim Relationship Validation
# --------------------------------------------------

print("Beneficiary-to-Claim Relationship Validation")
print("=" * 70)


# --------------------------------------------------
# Build beneficiary ID reference set
# --------------------------------------------------

beneficiary_ids = set()

for name in [
    "beneficiary_2008",
    "beneficiary_2009",
    "beneficiary_2010"
]:

    file_path = files[name]

    df = pd.read_csv(
        file_path,
        usecols=["DESYNPUF_ID"],
        low_memory=False
    )

    beneficiary_ids.update(
        df["DESYNPUF_ID"].dropna()
    )

print(f"Unique beneficiary IDs across annual files: {len(beneficiary_ids):,}")


# --------------------------------------------------
# Validate claim relationships
# --------------------------------------------------

for name in ["inpatient", "outpatient"]:

    file_path = files[name]

    total_claims = 0
    matched_claims = 0
    unmatched_claims = 0

    for chunk in pd.read_csv(
        file_path,
        usecols=["DESYNPUF_ID"],
        chunksize=100_000,
        low_memory=False
    ):

        claim_ids = chunk["DESYNPUF_ID"].dropna()

        total_claims += len(claim_ids)

        matched = claim_ids.isin(beneficiary_ids)

        matched_claims += matched.sum()
        unmatched_claims += (~matched).sum()

    match_rate = (
        matched_claims / total_claims * 100
        if total_claims > 0
        else 0
    )

    print(f"\n{name}")
    print(f"Total claims with beneficiary ID : {total_claims:,}")
    print(f"Matched beneficiary IDs          : {matched_claims:,}")
    print(f"Unmatched beneficiary IDs        : {unmatched_claims:,}")
    print(f"Match rate                       : {match_rate:.2f}%")

Beneficiary-to-Claim Relationship Validation
Unique beneficiary IDs across annual files: 116,352

inpatient
Total claims with beneficiary ID : 66,773
Matched beneficiary IDs          : 66,773
Unmatched beneficiary IDs        : 0
Match rate                       : 100.00%

outpatient
Total claims with beneficiary ID : 790,790
Matched beneficiary IDs          : 790,790
Unmatched beneficiary IDs        : 0
Match rate                       : 100.00%


## Beneficiary-to-Claim Relationship Findings

The relationship validation confirms that all claim records can be linked to a beneficiary ID contained in the beneficiary summary data.

### Results

| Dataset | Claims with Beneficiary ID | Matched | Unmatched | Match Rate |
|---|---:|---:|---:|---:|
| Inpatient | 66,773 | 66,773 | 0 | 100.00% |
| Outpatient | 790,790 | 790,790 | 0 | 100.00% |

### Beneficiary Population

The three annual beneficiary files contain a combined set of 116,352 unique `DESYNPUF_ID` values.

The annual files therefore represent repeated annual records for the same underlying beneficiary population rather than three completely separate populations.

### Data Quality Decision

No orphan claim records were identified.

The `DESYNPUF_ID` relationship between the claims datasets and beneficiary summary data is structurally complete for the records tested.

No claim records will be removed based on beneficiary linkage.

# Final Data Validation Summary

The initial structural validation of the CMS DE-SynPUF Sample 1 datasets has been completed.

## Dataset Availability

All five expected CSV datasets were successfully located and read:

- Beneficiary Summary 2008
- Inpatient Claims 2008–2010
- Outpatient Claims 2008–2010
- Beneficiary Summary 2009
- Beneficiary Summary 2010

## Record Counts

| Dataset | Records |
|---|---:|
| Beneficiary 2008 | 116,352 |
| Inpatient Claims | 66,773 |
| Outpatient Claims | 790,790 |
| Beneficiary 2009 | 114,538 |
| Beneficiary 2010 | 112,754 |

## Identifier Validation

Beneficiary `DESYNPUF_ID` values were unique within each annual beneficiary file.

Repeated `CLM_ID` values were identified in the claims datasets. However, validation of the composite `CLM_ID + SEGMENT` key showed:

- Inpatient: 0 duplicate combinations
- Outpatient: 0 duplicate combinations

Therefore, repeated `CLM_ID` values were not classified as erroneous duplicate records.

## Missing Value Validation

Missing values were identified across multiple fields.

Important patterns included:

- Approximately 98% missing `BENE_DEATH_DT`
- Several HCPCS/procedure fields with extremely high missingness
- Partial missingness in claim dates, diagnosis codes, HCPCS codes, and physician identifiers

Missing values will be evaluated based on field meaning and analytical relevance during the cleaning stage rather than being automatically removed or imputed.

## Data Type Validation

Key findings included:

- Identifiers are generally represented as strings or integer identifier fields.
- Financial fields such as `CLM_PMT_AMT` are numeric.
- Date fields are currently represented as numeric values and require conversion to datetime.
- Diagnosis and procedure code fields show mixed pandas data types and require explicit treatment during cleaning.

## Date Validation

No invalid or chronologically inconsistent dates were identified.

- No invalid beneficiary birth dates
- No invalid populated beneficiary death dates
- No death dates before birth dates
- No invalid claim start or end dates
- No claims with start dates later than end dates

## Financial and Numeric Validation

Negative and zero values were identified in several financial fields.

These values have not been classified as errors during validation.

Key findings include:

- Inpatient `CLM_PMT_AMT`: 55 negative values
- Outpatient `CLM_PMT_AMT`: 2,566 negative values
- Inpatient utilization days: no negative values
- Some beneficiary reimbursement fields contain negative values

Financial values will be retained until their analytical treatment is formally defined.

## Beneficiary-to-Claim Relationship

All claim records had a matching `DESYNPUF_ID` in the beneficiary reference population.

| Dataset | Match Rate |
|---|---:|
| Inpatient | 100.00% |
| Outpatient | 100.00% |

No orphan claim records were identified.

## Validation Conclusion

The datasets passed the major structural validation checks.

No source records were deleted or modified during this notebook.

The following areas require attention during the next stages:

1. High-missingness fields
2. Mixed data types in coded fields
3. Conversion of numeric date fields to datetime
4. Treatment and interpretation of negative financial values
5. Identification of fields that are analytically relevant to the project
6. Formal data-cleaning rules

### Data Handling Principle

> **Raw data remains unchanged.**

All transformations and cleaning decisions will be performed in subsequent notebooks and saved separately from the raw source files.